yaaredis/commands/iter.py

```diff
@@ -81,18 +84,32 @@ class ClusterIterCommandMixin(IterCommandMixin):
     async def scan_iter(self, match=None, count=None,
                         type=type):  # pylint: disable=redefined-builtin
         nodes = await self.cluster_nodes()
+
+        async def iterate_node(node, queue):
+            nonlocal match, count, type, nodes
+            cursor = '0'
+            while cursor != 0:
+                pieces = [cursor]
+                if match is not None:
+                    pieces.extend(['MATCH', match])
+                if count is not None:
+                    pieces.extend(['COUNT', count])
+                if type is not None:
+                    pieces.extend(['TYPE', type])
+                response = await self.execute_command_on_nodes(
+                    [node], 'SCAN', *pieces)
+                cursor, data = list(response.values())[0]
+                for item in data:
+                    await queue.put(item)  # blocks if queue is full
+
+        # maxsize ensures we don't pull too much data into
+        # memory if we are not processing it yet
+        queue = asyncio.Queue(maxsize=1000)
+        tasks = []
         for node in nodes:
             if 'master' in node['flags']:
-                cursor = '0'
-                while cursor != 0:
-                    pieces = [cursor]
-                    if match is not None:
-                        pieces.extend(['MATCH', match])
-                    if count is not None:
-                        pieces.extend(['COUNT', count])
-                    if type is not None:
-                        pieces.extend(['TYPE', type])
-                    response = await self.execute_command_on_nodes([node], 'SCAN', *pieces)
-                    cursor, data = list(response.values())[0]
-                    for item in data:
-                        yield item
+                t = asyncio.create_task(iterate_node(node, queue))
+                tasks.append(t)
+
+        while not all(t.done() for t in tasks):
```

Comment:
Is there a race condition here?
* assume a user has two nodes each with two keys
* we SCAN the first node and return a single item to the queue (`await queue.put(item)`)
* context returns here and we `yield await queue.get()` to the user
* the user's code does something like `for item in scan_iter(): await foo()` such that they preempt before the next iteration
* we SCAN the second item from node one and both items from node 2
* the user code finishes the iteration and we return to this method context
* we run `while not all(t.done() for t in tasks):` which is now false (since both tasks have completed)

I think we might need a final step of purging the queue in that case, eg. something like:
```python
while not all(t.done() for t in tasks):
    yield await queue.get()
while not queue.empty():
    yield await queue.get()
```

Although at that point, I suppose we could instead do something like this:
```python
while any(not t.done() for t in tasks) or not queue.empty():
    yield await queue.get()
```

Thoughts?

---

Alternatively, we could set the `asyncio.Queue(maxsize=1)` to guarantee that we return to the context of this method (and thus repeat the `while` check) between each operation; since two tasks cannot complete their final iteration in a single preemption loop in that case (ie. because one must block on the full queue), that would guarantee we avoid hitting the race condition as well.

This might also have the nice advantage of reducing memory usage -- though keeping 1k items in memory is _probably_ fine, keeping fewer items is always going to be safer.

Reply:
Ah yes I completely missed that case, this stuff is hard!

I'm happy with
`while any(not t.done() for t in tasks) or not queue.empty():`

The 1000 size of the buffer is fairly arbitrary; the idea is that the consumer shouldn't have to wait for an item (after waiting for the first SCAN call to return).   

I think a smaller buffer would work just as well, as the producers can happily be awaiting to populate the queue.  
So we could also set the maxsize to the `count` parameter (or 10 which is the redis-side default) on the basis that we'll have N nodes eagerly awaiting to populate the queue.  That would be good if you didn't want the SCAN on any given node proceeding ahead with the next iteration before it has been consumed.

I'm sure there's also a case for having `maxsize=((count or 10) * len(master_nodes)`.

Reply:
I've tested with maxsize=(count-1) and got no appreciable reduction in performance, so have included it as https://github.com/talkiq/yaaredis/pull/7/commits/83ecf75c428c19ad7116f54610198f844d815e19 (the choice of 9 rather than 10 is detailed in the commit).

> I'm sure there's also a case for having maxsize=((count or 10) * len(master_nodes).
Don't now think this is needed.